### PR TITLE
Use direct processes instead of shell invocations.

### LINKED ISSUE #273: Remove implicit dependency on shell
## What happened?

In [this Paketo RFC](https://github.com/paketo-buildpacks/rfcs/pull/136), there's discussion of the fact that the CNB project will soon remove the ability to run "indirect" processes – these will no longer run with a shell. This buildpack currently relies on a shell in 2 ways:
- the [standard process type is `Direct: false`](https://github.com/paketo-buildpacks/dotnet-execute/blob/8c26e75d941ce57ddbd4fa27c63b61dd1fba778f/build.go#L50-L56), meaning the launcher uses a shell to invoke it
- the [reloadable process type uses `--shell=sh`](https://github.com/paketo-buildpacks/dotnet-execute/blob/8c26e75d941ce57ddbd4fa27c63b61dd1fba778f/build.go#L66-L71) so that watchexec will invoke the start command with a shell

To remove implicit dependency on shell, the buildpack should move to explicitly invoking bash if needed for standard or reloadable processes.

## Checklist

* [ ] I have included log output.
* [ ] The log output includes an error message.
* [ ] I have included steps for reproduction.

### PR DESCRIPTION
**Pull request by joshuatcasey**
## Summary
Use direct processes instead of shell invocations. Requires a new helper binary to choose the port at launch time.

Closes #273.

Preserves the default port of `8080`.

## Non-Breaking Change
* Uses `ASPNETCORE_URLS` env variable instead of `--urls` argument
This uses environment variables (not command line args) to configure .NET core, which seems to be the K8s / 12 factor preferences. See References. If `PORT` is set, the port given will be set in `ASPNETCORE_URLS` at launch time, otherwise port `8080` is used.

## TODO
* [x] Refactor for readability.
* [x] Test the `helper` binary

## Context
* [Paketo RFC 41 `Use Direct Processes and exec.d` ](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0041-direct.md)
* [Buildpacks RFC 93 `Use Direct Processes and exec.d`](https://github.com/buildpacks/rfcs/blob/main/text/0093-remove-shell-processes.md)
* ["The twelve-factor app stores config in environment variables"](https://12factor.net/config)
* [.NET Core + Kestrel - how to set the port](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-6.0) (be sure to select the appropriate version!
* [https://andrewlock.net/5-ways-to-set-the-urls-for-an-aspnetcore-app/](https://andrewlock.net/5-ways-to-set-the-urls-for-an-aspnetcore-app/)

## Checklist
* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by sophiewigmore** [comment]
I think supporting `ASPNETCORE_URLS` is a good idea, but I don't think we should fully rip out support for `PORT` without giving some kind of warning to users that it'll go away (especially because it's a project-wide standard practice to use it more or less).

A new configuration option, to me, sounds like a cause for an RFC (albeit in this case a very small one), which would also serve as a record of the decision we make. To me it would be reasonable to support both environment variables, with `ASPNETCORE_URLS` taking precedence, as the .NET Core native choice. This could be something we lay out in the RFC.

@paketo-buildpacks/dotnet-core-maintainers  what do you think?

**Comment by sophiewigmore** [comment]
@joshuatcasey I appreciate all of the context links you've provided here for reference. That's super helpful

**Comment by sophiewigmore** [comment]
bump on this question @fg-j and @ForestEckhardt

**Comment by fg-j** [comment]
@sophiewigmore I agree about supporting `$PORT` and `$ASPNETCORE_URLS`. I'd be fine without an RFC if the change is backward compatible. This PR would be sufficient paper trail about the decision, in my opinion.

I think @joshuatcasey was intending to investigate using an `exec.d` script that could set the value of `$ASPNETCORE_URLS` based on `$PORT` to provide backward compatibility. That way, users could specify `$PORT` but the buildpack would always communicate with .NET using the  idiomatic`$ASPNETCORE_URLS` environment variable.

**Comment by joshuatcasey** [comment]
Hi team (@sophiewigmore @fg-j @ForestEckhardt), turns out I had github notifications turned off and missed all the discussion above. I plan to keep pushing this forward. I've enabled notifications should that shouldn't happen again.

For setting the port, it sounds like `PORT` is the unofficial Paketo standard so we don't want to deprecate it. I think we will need some kind of logic in an `exec.d` script or runner to replace the BASH interpolation, regardless of which env variable is used to set the port or how that's passed into the `dotnet` executable.

File: build_test.go
Comment (sophiewigmore) [inline build_test.go:131]: Not sure if I understand what this line is for. Wouldn't you want something like
`Expect(result.Layers[0].Path).To(Equal(layer.Path))`?

File: build_test.go
Comment (sophiewigmore) [inline build_test.go:33]: This is a nitpick, but can you name this `helperLayer` just for test clarity, and then also define the variable at the top and set the value in the `it.Before` below so it matches the other variables?

File: helper/helper.go
Comment (sophiewigmore) [inline helper/helper.go:12]: Can you add a comment about what this is doing for context!

File: scripts/build.sh
Comment (sophiewigmore) [inline scripts/build.sh:67]: @paketo-buildpacks/dotnet-core-maintainers will this get overridden by our "update github-config" workflows?

**Comment by sophiewigmore** [comment]
@joshuatcasey besides the small things I've pointed out, this looks pretty great to me!
@paketo-buildpacks/dotnet-core-maintainers I would love to get another set of eyes on this if possible since it's a heftier PR!

**Comment by ForestEckhardt** [comment]
So there is a canonical way of adding an execD process without having to modify the build script which you can see here https://github.com/paketo-buildpacks/node-engine/tree/main/cmd/optimize-memory. I would like it restructured in this format just to keep things straight forward. Also a change to the `build.sh` file will not survive a script sync.

File: build.go
Comment (ForestEckhardt) [inline build.go:132]: Generally speaking I am kinda of opposed to just having a private function like this. I see no reason why this shouldn't just be part of the main function in this format.

File: scripts/build.sh
Comment (fg-j) [reply scripts/build.sh:67]: yes, unless we add this file to the syncignore

Comment (joshuatcasey) [reply scripts/build.sh:67]: I think that https://github.com/paketo-buildpacks/dotnet-execute/pull/277#issuecomment-1064163463 has the solution, I will revert this change.

File: build.go
Comment (joshuatcasey) [reply build.go:132]: I'll inline it

File: build_test.go
Comment (joshuatcasey) [reply build_test.go:33]: done

File: helper/helper.go
Comment (joshuatcasey) [reply helper/helper.go:12]: done

File: helper/helper.go
Comment (joshuatcasey) [reply helper/helper.go:28]: It should look like the below. It's logged on app launch.

`Setting ASPNETCORE_URLS=http://0.0.0.0:8080/`

File: build_test.go
Comment (joshuatcasey) [reply build_test.go]: I would prefer that, but it looks like the `layer.Path` is autogenerated and I don't think there's a way to check equality. I've (slightly) modified how this is done and the comment that goes with it, could you take another look?

File: build_test.go
Comment (sophiewigmore) [inline build_test.go:134]: Since you've set up `layersDir` and that's used in the build you can do:

```
Expect(result.Layers[0].Path).To(Equal(filepath.Join(layersDir, "port-chooser")))
```

File: build_test.go
Comment (sophiewigmore) [inline build_test.go:54]: This can be:
` Path:             filepath.Join(layersDir, "port-chooser")`

and then wherever you are doing `portChooserLayer.Path = result.Layers[0].Path` can be removed.

File: helper/helper.go
Comment (sophiewigmore) [reply helper/helper.go]: Sorry, I rather confusingly only highlighted this part, but I meant for the entire file (now cmd/port-chooser/internal/port-chooser.go`. Just a one liner explaining the port selection would be sufficient.

File: build_test.go
Comment (joshuatcasey) [reply build_test.go:54]: Done, good idea

Comment (joshuatcasey) [reply build_test.go:54]: done, will push in a minute

**Review by sophiewigmore — APPROVED** [review]

**Pull request merged**
